File: sbt-bridge/src/xsbt/ConsoleInterface.scala
/*
 * Dotty (https://dotty.epfl.ch/)
 *
 * Copyright EPFL.
 *
 * Licensed under Apache License 2.0
 * (https://www.apache.org/licenses/LICENSE-2.0).
 */

package xsbt

import xsbti.Logger

import dotty.tools.dotc.core.Contexts.Context
import dotty.tools.repl.ReplDriver

class ConsoleInterface {
  def commandArguments(
    args: Array[String],
    bootClasspathString: String,
    classpathString: String,
    log: Logger
  ): Array[String] = args

  def run(args: Array[String],
          bootClasspathString: String,
          classpathString: String,
          initialCommands: String,
          cleanupCommands: String,
          loader: ClassLoader,
          bindNames: Array[String],
          bindValues: Array[Any],
          log: Logger
  ): Unit = {
    val completeArgs =
      args ++ {
        if (bootClasspathString.isEmpty) Array.empty[String]
        else Array("-bootclasspath", bootClasspathString)
      } ++
      Array("-classpath", classpathString)

    val driver = new ReplDriver(completeArgs, classLoader = Some(loader))

    val s0 = (bindNames, bindValues).zipped.foldLeft(driver.initialState) {
      case (state, (name, value)) => driver.bind(name, value)(state)
    }

    val s1 = driver.run(initialCommands)(s0)
    // TODO handle failure during initialisation
    val s2 = driver.runUntilQuit(s1)
    driver.run(cleanupCommands)(s2)
  }
}

File: sbt-bridge/src-bootstrapped/ConsoleInterface.java
/* sbt -- Simple Build Tool
 * Copyright 2008, 2009 Mark Harrah
 */
package xsbt;

import java.util.ArrayList;

import scala.Option;

import xsbti.Logger;

import dotty.tools.dotc.core.Contexts.Context;
import dotty.tools.repl.ReplDriver;
import dotty.tools.repl.State;

public class ConsoleInterface {
  public String[] commandArguments(String[] args, String bootClasspathString, String classpathString, Logger log) {
    return args;
  }

  public void run(
    String[] args,
    String bootClasspathString,
    String classpathString,
    String initialCommands,
    String cleanupCommands,
    ClassLoader loader,
    String[] bindNames,
    Object[] bindValues,
    Logger log
  ) {
    ArrayList<String> completeArgsList = new ArrayList<>();
    for (String arg : args)
      completeArgsList.add(arg);
    if (!bootClasspathString.isEmpty()) {
      completeArgsList.add("-bootclasspath");
      completeArgsList.add(bootClasspathString);
    }
    completeArgsList.add("-classpath");
    completeArgsList.add(classpathString);
    String[] completeArgs = completeArgsList.toArray(args);

    ReplDriver driver = new ReplDriver(completeArgs, System.out, Option.apply(loader), ReplDriver.pprintImport());

    State state = driver.initialState();
    assert bindNames.length == bindValues.length;
    for (int i = 0; i < bindNames.length; i++)
      state = driver.bind(bindNames[i], bindValues[i], state);

    state = driver.run(initialCommands, state);
    // TODO handle failure during initialisation
    state = driver.runUntilQuit(state);
    driver.run(cleanupCommands, state);
  }
}
